raiden_contracts/contract_manager.py

```diff
@@ -85,77 +91,99 @@ def _compile_all_contracts(self) -> None:
+            raise ContractManagerCompilationError("Checksumming failed.")
+
+        if not self.contracts:
             self._compile_all_contracts()
 
         target_path.parent.mkdir(parents=True, exist_ok=True)
-        with gzip.GzipFile(target_path, 'wb') as target_file:
+        with target_path.open(mode='w') as target_file:
             target_file.write(
                 json.dumps(
-                    dict(contracts=self._contracts, checksum=contracts_checksum),
-                ).encode(),
+                    dict(
+                        contracts=self.contracts,
+                        contracts_checksums=self.contracts_checksums,
+                        overall_checksum=self.overall_checksum,
+                    ),
+                ),
             )
 
     def get_contract(self, contract_name: str) -> Dict:
         """ Return ABI, BIN of the given contract. """
-        if not self._contracts:
+        if not self.contracts:
             self._compile_all_contracts()
-        return self._contracts[contract_name]
+        return self.contracts[contract_name]
 
     def get_contract_abi(self, contract_name: str) -> Dict:
         """ Returns the ABI for a given contract. """
-        if not self._contracts:
+        if not self.contracts:
             self._compile_all_contracts()
-        return self._contracts[contract_name]['abi']
+        return self.contracts[contract_name]['abi']
 
     def get_event_abi(self, contract_name: str, event_name: str) -> Dict:
         """ Returns the ABI for a given event. """
         # Import locally to avoid web3 dependency during installation via `compile_contracts`
         from web3.utils.contracts import find_matching_event_abi
 
-        if not self._contracts:
+        if not self.contracts:
             self._compile_all_contracts()
         contract_abi = self.get_contract_abi(contract_name)
         return find_matching_event_abi(contract_abi, event_name)
 
-    def _checksum_contracts(self):
-        if self._contracts_source_dirs is None:
-            raise TypeError("Can't checksum when using precompiled contracts.")
-        checksums = []
-        for contracts_dir in self._contracts_source_dirs.values():
+    def checksum_contracts(self) -> None:
+        if self.contracts_source_dirs is None:
+            raise TypeError("Missing contracts source path, can't checksum contracts.")
+
+        checksums = {}
+        for contracts_dir in self.contracts_source_dirs.values():
             file: Path
             for file in contracts_dir.glob('*.sol'):
-                checksums.append(
-                    '{}:{}'.format(
-                        file.name,
-                        hashlib.sha256(file.read_bytes()).hexdigest(),
-                    ),
+                checksums[file.name] = hashlib.sha256(file.read_bytes()).hexdigest()
+
+        self.overall_checksum = hashlib.sha256(
+            ':'.join([checksums[key] for key in sorted(checksums)]).encode(),
```

Comment:
btw, join accepts an iterable, and `(.. for ... in ...)` is a generator, which is an iterable, so the inner list is unnecessary, the following has the same effect and does one allocation less.

> ':'.join(checksums[key] for key in sorted(checksums)).encode()